### PR TITLE
Fix ddp gradient view slice

### PR DESCRIPTION
**Pull request by Flowingsun007**
- [x] SliceView1dContiguous通过view::IsViewApplicable做check，否则当条件不满足（譬如设置了环境变量`ONEFLOW_DISABLE_VIEW=true`时，ddp下训练会报错）
- [x] 增加python env util，处理环境变量
- [x] tensor.cuda修复bug并增加test case

### CODE REVIEW

**Comment by daquexian** [comment]
这里的 view 是不能关掉的不然结果会不对，如果真的不能使用 view，就报一个错提示用户该把 DistributedDataParallel 的 bucket_size 参数设成 1 吧

**Comment by Flowingsun007** [comment]
> 这里的 view 是不能关掉的不然结果会不对，如果真的不能使用 view，就报一个错提示用户该把 DistributedDataParallel 的 bucket_size 参数设成 1 吧

已修改

File: python/oneflow/nn/parallel/ddp.py [hunk 1]
```diff
@@ -67,7 +69,9 @@ def DistributedDataParallel(
     module: "flow.nn.Module", *, broadcast_buffers: bool = True, bucket_size: int = 10
 ):
     assert all(x.dtype == flow.float32 for x in module.parameters())
-
+    if(os.getenv('ONEFLOW_DESABLE_VIEW') in ('true', 'True')):
```
Comment (hjchen2) [inline python/oneflow/nn/parallel/ddp.py:72]: 建议改为`if not os.getenv('ONEFLOW_DESABLE_VIEW') in ('false', 'False', '0', 0)`

Comment (Flowingsun007) [reply python/oneflow/nn/parallel/ddp.py:72]: 好建议

Comment (Flowingsun007) [reply python/oneflow/nn/parallel/ddp.py]: 已修改

File: python/oneflow/nn/parallel/ddp.py [hunk 2]
```diff
@@ -67,7 +69,9 @@ def DistributedDataParallel(
     module: "flow.nn.Module", *, broadcast_buffers: bool = True, bucket_size: int = 10
 ):
     assert all(x.dtype == flow.float32 for x in module.parameters())
-
+    if not os.getenv('ONEFLOW_DESABLE_VIEW') in ('false', 'False', '0', 0):
```
Comment (daquexian) [inline python/oneflow/nn/parallel/ddp.py:72]: 实现一个公共函数吧，仿照 oneflow/core/common/util.cpp:145，以及环境变量的值只会是字符串

Comment (Flowingsun007) [reply python/oneflow/nn/parallel/ddp.py]: 已增加

File: python/oneflow/nn/parallel/ddp.py [hunk 3]
```diff
@@ -67,7 +80,9 @@ def DistributedDataParallel(
     module: "flow.nn.Module", *, broadcast_buffers: bool = True, bucket_size: int = 10
 ):
     assert all(x.dtype == flow.float32 for x in module.parameters())
-
+    if parse_boolean_form_env("ONEFLOW_DESABLE_VIEW", True):
+        warnings.warn("because the environment variable 'ONEFLOW_DESABLE_VIEW' is set to true, so the view mechanism is disabled, and we will set bucket_size = 1")
```
Comment (BBuf) [inline python/oneflow/nn/parallel/ddp.py:84]: 我比较好奇这个环境变量为什么叫 DESABLE ？而不是DISABLE呢

Comment (Flowingsun007) [reply python/oneflow/nn/parallel/ddp.py]: 已修改

**Review by BBuf — APPROVED** [review]

File: python/oneflow/nn/parallel/ddp.py [hunk 4]
```diff
@@ -67,7 +80,9 @@ def DistributedDataParallel(
     module: "flow.nn.Module", *, broadcast_buffers: bool = True, bucket_size: int = 10
 ):
     assert all(x.dtype == flow.float32 for x in module.parameters())
-
+    if parse_boolean_form_env("ONEFLOW_DISABLE_VIEW", True):
```
Comment (daquexian) [inline python/oneflow/nn/parallel/ddp.py:83]:
```suggestion
    if parse_boolean_from_env("ONEFLOW_DISABLE_VIEW", True):
```

File: python/oneflow/nn/parallel/ddp.py [hunk 5]
```diff
@@ -13,12 +13,25 @@
...
 """
+import os
+import warnings
 from collections import OrderedDict
 
 import oneflow as flow
 from oneflow.framework.tensor_tuple_util import convert_to_tensor_tuple
 
 
+def parse_boolean_form_env(string, flag):
```
Comment (daquexian) [inline python/oneflow/nn/parallel/ddp.py:24]: 公共函数，放在一个公共文件里吧，可以新建一个 python/oneflow/support/env_var_util.py

File: python/oneflow/nn/parallel/ddp.py [hunk 6]
```diff
@@ -13,12 +13,25 @@
...
+    assert isinstance(string, str), "env variable must be string, but got: " + type(string)
+    assert isinstance(flag, bool), "env variable flag must be boolean, but got: " + type(flag)
+    if(flag):
+        if os.getenv(string) in ('True', 'true', '1', 1):
```
Comment (daquexian) [inline python/oneflow/nn/parallel/ddp.py:28]: 环境变量只有字符串，不会有数字 1 和 0 的。

可以参考 oneflow/core/common/util.cpp 的 ParseBooleanFromEnv。现在不能正确识别 TRUE 和 ON/on

File: python/oneflow/nn/parallel/ddp.py (hunk 5 above)
Comment (daquexian) [reply python/oneflow/nn/parallel/ddp.py:24]: 参数命名可以修改的更合理

File: python/oneflow/nn/parallel/ddp.py [hunk 7]
```diff
@@ -13,12 +13,25 @@
...
 from collections import OrderedDict
 
 import oneflow as flow
 from oneflow.framework.tensor_tuple_util import convert_to_tensor_tuple
 
 
+def parse_boolean_form_env(string, flag):
+    assert isinstance(string, str), "env variable must be string, but got: " + type(string)
+    assert isinstance(flag, bool), "env variable flag must be boolean, but got: " + type(flag)
+    if(flag):
```
Comment (daquexian) [inline python/oneflow/nn/parallel/ddp.py:27]: 这个逻辑和 c++ 的 ParseBooleanFromEnv 没有对齐。ParseBooleanFromEnv 里第二个参数是默认值

File: python/oneflow/nn/parallel/ddp.py (hunk 6 above)
Comment (daquexian) [reply python/oneflow/nn/parallel/ddp.py:28]: 也可以考虑直接将 c++ 的 ParseBooleanFromEnv 导出到 Python 层

**Review by daquexian — APPROVED** [review]

File: python/oneflow/nn/parallel/ddp.py (hunk 7 above)
Comment (Flowingsun007) [reply python/oneflow/nn/parallel/ddp.py]: 上述已修改
